# Homework 02

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("videosUS.csv")

In [3]:
# How the data looks like
df.head(3)

video_id trending_date  \
0  2kyS6SvSYSE      17.14.11   
1  1ZAPwfrtAFY      17.14.11   
2  5qpjK5DgCt4      17.14.11   

                                               title    channel_title  \
0                 WE WANT TO TALK ABOUT OUR MARRIAGE     CaseyNeistat   
1  The Trump Presidency: Last Week Tonight with J...  LastWeekTonight   
2  Racist Superman | Rudy Mancuso, King Bach & Le...     Rudy Mancuso   

   category_id              publish_time  \
0         22.0  2017-11-13T17:13:01.000Z   
1         24.0  2017-11-13T07:30:00.000Z   
2         23.0  2017-11-12T19:05:24.000Z   

                                                tags      views     likes  \
0                                    SHANtell martin   748374.0   57527.0   
1  last week tonight trump presidency|"last week ...  2418783.0   97185.0   
2  racist superman|"rudy"|"mancuso"|"king"|"bach"...  3191434.0  146033.0   

   dislikes  comment_count                                  thumbnail_link  \
0    2966.0        15954.0  https://i.ytimg.com/vi/2kyS6SvSYSE/default.jpg   
1    6146.0        12703.0  https://i.ytimg.com/vi/1ZAPwfrtAFY/default.jpg   
2    5339.0         8181.0  https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg   

  comments_disabled ratings_disabled video_error_or_removed  \
0             False            False                  False   
1             False            False                  False   
2             False            False                  False   

                                         description  
0  SHANTELL'S CHANNEL - https://www.youtube.com/s...  
1  One year after the presidential election, John...  
2  WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...

In [4]:
len(df)

41415

### How many missing values within the dataset 

#### In each column

In [5]:
df.isna().sum()

video_id                     0
trending_date              466
title                      466
channel_title              466
category_id                466
publish_time               466
tags                       466
views                      466
likes                      466
dislikes                   466
comment_count              466
thumbnail_link             466
comments_disabled          466
ratings_disabled           466
video_error_or_removed     466
description               1036
dtype: int64

In [6]:
df[df.title.isna()]

video_id trending_date title  \
53     \nSubscribe to WWE on YouTube: http://bit.ly/1...           NaN   NaN   
54                 \nVisit WWE.com: http://goo.gl/akf0J4           NaN   NaN   
55     \nMust-See WWE videos on YouTube: https://goo....           NaN   NaN   
319    \nSubscribe to WWE on YouTube: http://bit.ly/1...           NaN   NaN   
320                \nVisit WWE.com: http://goo.gl/akf0J4           NaN   NaN   
...                                                  ...           ...   ...   
36570  Follow me on Twitter ► https://twitter.com/mar...           NaN   NaN   
36571  Like me on Facebook ► https://www.facebook.com...           NaN   NaN   
36572  Join us on Reddit! ► https://www.reddit.com/r/...           NaN   NaN   
36573  Horror Outro ► https://soundcloud.com/shurkoff...           NaN   NaN   
36574  Happy Outro ► https://soundcloud.com/hielia/mi...           NaN   NaN   

      channel_title  category_id publish_time tags  views  likes  dislikes  \
53              NaN          NaN          NaN  NaN    NaN    NaN       NaN   
54              NaN          NaN          NaN  NaN    NaN    NaN       NaN   
55              NaN          NaN          NaN  NaN    NaN    NaN       NaN   
319             NaN          NaN          NaN  NaN    NaN    NaN       NaN   
320             NaN          NaN          NaN  NaN    NaN    NaN       NaN   
...             ...          ...          ...  ...    ...    ...       ...   
36570           NaN          NaN          NaN  NaN    NaN    NaN       NaN   
36571           NaN          NaN          NaN  NaN    NaN    NaN       NaN   
36572           NaN          NaN          NaN  NaN    NaN    NaN       NaN   
36573           NaN          NaN          NaN  NaN    NaN    NaN       NaN   
36574           NaN          NaN          NaN  NaN    NaN    NaN       NaN   

       comment_count thumbnail_link comments_disabled ratings_disabled  \
53               NaN            NaN               NaN              NaN   
54               NaN            NaN               NaN              NaN   
55               NaN            NaN               NaN              NaN   
319              NaN            NaN               NaN              NaN   
320              NaN            NaN               NaN              NaN   
...              ...            ...               ...              ...   
36570            NaN            NaN               NaN              NaN   
36571            NaN            NaN               NaN              NaN   
36572            NaN            NaN               NaN              NaN   
36573            NaN            NaN               NaN              NaN   
36574            NaN            NaN               NaN              NaN   

      video_error_or_removed description  
53                       NaN         NaN  
54                       NaN         NaN  
55                       NaN         NaN  
319                      NaN         NaN  
320                      NaN         NaN  
...                      ...         ...  
36570                    NaN         NaN  
36571                    NaN         NaN  
36572                    NaN         NaN  
36573                    NaN         NaN  
36574                    NaN         NaN  

[466 rows x 16 columns]

#### My assumption about why we have these missing values

Looking at the previous two results, we can clearly see that there's 466 missing value in each column (except the description, and video_id columns), and actually, these 466 missing values are common to the same videos. Maybe it means that these videos were deleted or can't be shared on public. Therefore, we can delete them.
The description column though, has 1036 missing values, which is pretty normal I guess, since descriptions are usually optional.

#### Getting rid of the 466 columns

In [7]:
df.drop(df[(df.title.isna())].index, inplace=True)

In [8]:
df[df.title.isna()]

Empty DataFrame
Columns: [video_id, trending_date, title, channel_title, category_id, publish_time, tags, views, likes, dislikes, comment_count, thumbnail_link, comments_disabled, ratings_disabled, video_error_or_removed, description]
Index: []

### Mean , Median , and Quartiles of each numerical variable within this dataset.

In [9]:
columns = ["views", "likes", "dislikes", "comment_count"]

#### Mean

In [10]:
for column in columns:
    print("Mean of", column, ":", df[column].mean())

Mean of views : 2360784.6382573447
Mean of likes : 74266.7024347359
Mean of dislikes : 3711.400888910596
Mean of comment_count : 8446.803682629612


#### Median

In [11]:
for column in columns:
    print("Median of", column, ":", df[column].median())

Median of views : 681861.0
Median of likes : 18091.0
Median of dislikes : 631.0
Median of comment_count : 1856.0


#### Quartiles

In [12]:
for column in columns:
    print("Quartiles of", column)
    print(" 25% :", df[column].quantile(0.25))
    print(" 50% :", df[column].quantile(0.50))
    print(" 75% :", df[column].quantile(0.75))
    print("\n")

Quartiles of views
 25% : 242329.0
 50% : 681861.0
 75% : 1823157.0


Quartiles of likes
 25% : 5424.0
 50% : 18091.0
 75% : 55417.0


Quartiles of dislikes
 25% : 202.0
 50% : 631.0
 75% : 1938.0


Quartiles of comment_count
 25% : 614.0
 50% : 1856.0
 75% : 5755.0




I could have used df.describe() but I just wanted to display only Mean, Mediane, and Quartiles, and I didn't want to take the category_id into consideration, I don't know if it's good or not.

### Spot the possible outliers in your dataset and get rid of them.

#### Calculating the z-scores

In [13]:
df["views_zscore"] = (df.views - df.views.mean()) / df.views.std()
df["likes_zscore"] = (df.likes - df.likes.mean()) / df.likes.std()
df["dislikes_zscore"] = (df.dislikes - df.dislikes.mean()) / df.dislikes.std()
df["comment_count_zscore"] = (df.comment_count - df.comment_count.mean()) / df.comment_count.std()

In [14]:
# Now I will keep only the rows that doesn't contain outliers
# I will remove all the outliers from all the numerical rows
df = df[abs(df["views_zscore"]) < 3]
df = df[abs(df["likes_zscore"]) < 3]
df = df[abs(df["dislikes_zscore"]) < 3]
df = df[abs(df["comment_count_zscore"]) < 3]

In [15]:
# Dropping the columns zscore because we don't need them
df.drop(['views_zscore', 'likes_zscore', 'dislikes_zscore', 'comment_count_zscore'], axis = 1, inplace = True)

In [16]:
len(df)

40086

### How many unique categories we have

In [17]:
print("Unique categories:",df["category_id"].unique())
print("Number of unique categories:",len(df["category_id"].unique()))

Unique categories: [22. 24. 23. 28.  1. 25. 17. 10. 15. 27. 26.  2. 19. 20. 29. 43.]
Number of unique categories: 16


### Is there any column that we need to adjust its type

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 40086 entries, 0 to 41413
Data columns (total 16 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   video_id                40086 non-null  object 
 1   trending_date           40086 non-null  object 
 2   title                   40086 non-null  object 
 3   channel_title           40086 non-null  object 
 4   category_id             40086 non-null  float64
 5   publish_time            40086 non-null  object 
 6   tags                    40086 non-null  object 
 7   views                   40086 non-null  float64
 8   likes                   40086 non-null  float64
 9   dislikes                40086 non-null  float64
 10  comment_count           40086 non-null  float64
 11  thumbnail_link          40086 non-null  object 
 12  comments_disabled       40086 non-null  object 
 13  ratings_disabled        40086 non-null  object 
 14  video_error_or_removed  40086 non-null

In [19]:
df.head(10)

video_id trending_date  \
0  2kyS6SvSYSE      17.14.11   
1  1ZAPwfrtAFY      17.14.11   
2  5qpjK5DgCt4      17.14.11   
3  puqaWrEC7tY      17.14.11   
4  d380meD0W0M      17.14.11   
5  gHZ1Qz0KiKM      17.14.11   
6  39idVpFF7NQ      17.14.11   
7  nc99ccSXST0      17.14.11   
8  jr9QtXwC9vc      17.14.11   
9  TUmyygCMMGA      17.14.11   

                                               title          channel_title  \
0                 WE WANT TO TALK ABOUT OUR MARRIAGE           CaseyNeistat   
1  The Trump Presidency: Last Week Tonight with J...        LastWeekTonight   
2  Racist Superman | Rudy Mancuso, King Bach & Le...           Rudy Mancuso   
3                   Nickelback Lyrics: Real or Fake?  Good Mythical Morning   
4                           I Dare You: GOING BALD!?               nigahiga   
5                              2 Weeks with iPhone X               iJustine   
6          Roy Moore & Jeff Sessions Cold Open - SNL    Saturday Night Live   
7                5 Ice Cream Gadgets put to the Test     CrazyRussianHacker   
8  The Greatest Showman | Official Trailer 2 [HD]...       20th Century Fox   
9  Why the rise of the robots won’t mean the end ...                    Vox   

   category_id              publish_time  \
0         22.0  2017-11-13T17:13:01.000Z   
1         24.0  2017-11-13T07:30:00.000Z   
2         23.0  2017-11-12T19:05:24.000Z   
3         24.0  2017-11-13T11:00:04.000Z   
4         24.0  2017-11-12T18:01:41.000Z   
5         28.0  2017-11-13T19:07:23.000Z   
6         24.0  2017-11-12T05:37:17.000Z   
7         28.0  2017-11-12T21:50:37.000Z   
8          1.0  2017-11-13T14:00:23.000Z   
9         25.0  2017-11-13T13:45:16.000Z   

                                                tags      views     likes  \
0                                    SHANtell martin   748374.0   57527.0   
1  last week tonight trump presidency|"last week ...  2418783.0   97185.0   
2  racist superman|"rudy"|"mancuso"|"king"|"bach"...  3191434.0  146033.0   
3  rhett and link|"gmm"|"good mythical morning"|"...   343168.0   10172.0   
4  ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...  2095731.0  132235.0   
5  ijustine|"week with iPhone X"|"iphone x"|"appl...   119180.0    9763.0   
6  SNL|"Saturday Night Live"|"SNL Season 43"|"Epi...  2103417.0   15993.0   
7  5 Ice Cream Gadgets|"Ice Cream"|"Cream Sandwic...   817732.0   23663.0   
8  Trailer|"Hugh Jackman"|"Michelle Williams"|"Za...   826059.0    3543.0   
9  vox.com|"vox"|"explain"|"shift change"|"future...   256426.0   12654.0   

   dislikes  comment_count                                  thumbnail_link  \
0    2966.0        15954.0  https://i.ytimg.com/vi/2kyS6SvSYSE/default.jpg   
1    6146.0        12703.0  https://i.ytimg.com/vi/1ZAPwfrtAFY/default.jpg   
2    5339.0         8181.0  https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg   
3     666.0         2146.0  https://i.ytimg.com/vi/puqaWrEC7tY/default.jpg   
4    1989.0        17518.0  https://i.ytimg.com/vi/d380meD0W0M/default.jpg   
5     511.0         1434.0  https://i.ytimg.com/vi/gHZ1Qz0KiKM/default.jpg   
6    2445.0         1970.0  https://i.ytimg.com/vi/39idVpFF7NQ/default.jpg   
7     778.0         3432.0  https://i.ytimg.com/vi/nc99ccSXST0/default.jpg   
8     119.0          340.0  https://i.ytimg.com/vi/jr9QtXwC9vc/default.jpg   
9    1363.0         2368.0  https://i.ytimg.com/vi/TUmyygCMMGA/default.jpg   

  comments_disabled ratings_disabled video_error_or_removed  \
0             False            False                  False   
1             False            False                  False   
2             False            False                  False   
3             False            False                  False   
4             False            False                  False   
5             False            False                  False   
6             False            False                  False   
7             False            False                  False   
8             False     

There are some columns that we need to adjust their types :
- trending_date and publish_time must have the type of datetime64.
- category_id must have the type of int64 or category.
- views, likes, dislikes, comment_count must have the type of int64.
- comments_disabled, ratings_disabled, video_error_or_removed must have the type of  bool.

In [27]:
# Let's change the types
# I had trouble converting the date tho
convert_dict = {
                'category_id': int,
                'views': int,
                'likes': int,
                'dislikes': int,
                'comment_count': int,
                'comments_disabled': bool,
                'ratings_disabled': bool,
                'video_error_or_removed':bool
               }
  
df = df.astype(convert_dict)

In [28]:
df.head(10)

video_id trending_date  \
0  2kyS6SvSYSE      17.14.11   
1  1ZAPwfrtAFY      17.14.11   
2  5qpjK5DgCt4      17.14.11   
3  puqaWrEC7tY      17.14.11   
4  d380meD0W0M      17.14.11   
5  gHZ1Qz0KiKM      17.14.11   
6  39idVpFF7NQ      17.14.11   
7  nc99ccSXST0      17.14.11   
8  jr9QtXwC9vc      17.14.11   
9  TUmyygCMMGA      17.14.11   

                                               title          channel_title  \
0                 WE WANT TO TALK ABOUT OUR MARRIAGE           CaseyNeistat   
1  The Trump Presidency: Last Week Tonight with J...        LastWeekTonight   
2  Racist Superman | Rudy Mancuso, King Bach & Le...           Rudy Mancuso   
3                   Nickelback Lyrics: Real or Fake?  Good Mythical Morning   
4                           I Dare You: GOING BALD!?               nigahiga   
5                              2 Weeks with iPhone X               iJustine   
6          Roy Moore & Jeff Sessions Cold Open - SNL    Saturday Night Live   
7                5 Ice Cream Gadgets put to the Test     CrazyRussianHacker   
8  The Greatest Showman | Official Trailer 2 [HD]...       20th Century Fox   
9  Why the rise of the robots won’t mean the end ...                    Vox   

   category_id              publish_time  \
0           22  2017-11-13T17:13:01.000Z   
1           24  2017-11-13T07:30:00.000Z   
2           23  2017-11-12T19:05:24.000Z   
3           24  2017-11-13T11:00:04.000Z   
4           24  2017-11-12T18:01:41.000Z   
5           28  2017-11-13T19:07:23.000Z   
6           24  2017-11-12T05:37:17.000Z   
7           28  2017-11-12T21:50:37.000Z   
8            1  2017-11-13T14:00:23.000Z   
9           25  2017-11-13T13:45:16.000Z   

                                                tags    views   likes  \
0                                    SHANtell martin   748374   57527   
1  last week tonight trump presidency|"last week ...  2418783   97185   
2  racist superman|"rudy"|"mancuso"|"king"|"bach"...  3191434  146033   
3  rhett and link|"gmm"|"good mythical morning"|"...   343168   10172   
4  ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...  2095731  132235   
5  ijustine|"week with iPhone X"|"iphone x"|"appl...   119180    9763   
6  SNL|"Saturday Night Live"|"SNL Season 43"|"Epi...  2103417   15993   
7  5 Ice Cream Gadgets|"Ice Cream"|"Cream Sandwic...   817732   23663   
8  Trailer|"Hugh Jackman"|"Michelle Williams"|"Za...   826059    3543   
9  vox.com|"vox"|"explain"|"shift change"|"future...   256426   12654   

   dislikes  comment_count                                  thumbnail_link  \
0      2966          15954  https://i.ytimg.com/vi/2kyS6SvSYSE/default.jpg   
1      6146          12703  https://i.ytimg.com/vi/1ZAPwfrtAFY/default.jpg   
2      5339           8181  https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg   
3       666           2146  https://i.ytimg.com/vi/puqaWrEC7tY/default.jpg   
4      1989          17518  https://i.ytimg.com/vi/d380meD0W0M/default.jpg   
5       511           1434  https://i.ytimg.com/vi/gHZ1Qz0KiKM/default.jpg   
6      2445           1970  https://i.ytimg.com/vi/39idVpFF7NQ/default.jpg   
7       778           3432  https://i.ytimg.com/vi/nc99ccSXST0/default.jpg   
8       119            340  https://i.ytimg.com/vi/jr9QtXwC9vc/default.jpg   
9      1363           2368  https://i.ytimg.com/vi/TUmyygCMMGA/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   
2              False             False                   False   
3              False             False                   False   
4              False             False                   False   
5              False             False                   False   
6              False             False                   False   
7              False             False                   False   
8              False             False   

### How many tags we have and which are the most common tags

In [42]:
# So each video has a bunch of tags
df.tags.unique()

array(['SHANtell martin',
       'last week tonight trump presidency|"last week tonight donald trump"|"john oliver trump"|"donald trump"',
       'racist superman|"rudy"|"mancuso"|"king"|"bach"|"racist"|"superman"|"love"|"rudy mancuso poo bear black white official music video"|"iphone x by pineapple"|"lelepons"|"hannahstocking"|"rudymancuso"|"inanna"|"anwar"|"sarkis"|"shots"|"shotsstudios"|"alesso"|"anitta"|"brazil"|"Getting My Driver\'s License | Lele Pons"',
       ...,
       'The Late Late Show|"Late Late Show"|"James Corden"|"Corden"|"late night"|"late night show"|"comedy"|"comedian"|"celebrity"|"celeb"|"celebrities"|"CBS"|"joke"|"jokes"|"funny"|"funny videos"|"funny video"|"humor"|"hollywood"|"famous"|"Drop the Mic"|"Ashton Kutcher"|"Sean Diddy Combs"|"Segment"|"Game"|"Music"|"Recurring"|"Evergreen"',
       'bon appetit|"burgers"|"cheeseburgers"|"how to cook"|"test kitchen"|"cast iron"|"smashburgers"|"how to make cheeseburgers"|"how to make smashburgers"|"how to make a smashburg

In [43]:
len(df.tags)

40086

In [44]:
len(df.tags.unique())

6020

In [45]:
# Extracting all the tags that we have
tags_collection = df.tags.unique()
tags = []
for i in range(0,len(tags_collection)):
    tags.extend(tags_collection[i].replace('"','').split("|"))

In [39]:
tags

['SHANtell martin',
 'last week tonight trump presidency',
 'last week tonight donald trump',
 'john oliver trump',
 'donald trump',
 'racist superman',
 'rudy',
 'mancuso',
 'king',
 'bach',
 'racist',
 'superman',
 'love',
 'rudy mancuso poo bear black white official music video',
 'iphone x by pineapple',
 'lelepons',
 'hannahstocking',
 'rudymancuso',
 'inanna',
 'anwar',
 'sarkis',
 'shots',
 'shotsstudios',
 'alesso',
 'anitta',
 'brazil',
 "Getting My Driver's License ",
 ' Lele Pons',
 'rhett and link',
 'gmm',
 'good mythical morning',
 'rhett and link good mythical morning',
 'good mythical morning rhett and link',
 'mythical morning',
 'Season 12',
 'nickelback lyrics',
 'nickelback lyrics real or fake',
 'nickelback',
 'nickelback songs',
 'nickelback song',
 'rhett link nickelback',
 'gmm nickelback',
 'lyrics (website category)',
 'nickelback (musical group)',
 'rock',
 'music',
 'lyrics',
 'chad kroeger',
 'canada',
 'music (industry)',
 'mythical',
 'gmm challenge',
 'c

In [40]:
len(tags)

125969

In [46]:
# Getting only the unique tags
unique = []
for tag in tags:
    if tag not in unique:
        unique.append(tag)

In [47]:
len(unique)

56449

### Top 10 Tags

In [48]:
from collections import Counter

tags = df.tags.values
all_tags = []
for i in range(0,len(tags)):
    all_tags.extend(tags[i].replace('"','').split("|"))

Counter(all_tags).most_common(10)

[('funny', 3582),
 ('comedy', 2915),
 ('how to', 1586),
 ('[none]', 1535),
 ('music', 1302),
 ('2018', 1270),
 ('humor', 1185),
 ('food', 1152),
 ('Pop', 1132),
 ('science', 1111)]